man/man3/TraceLog.3
.TH TraceLog 3 "Sep 2012" "TraceLog" "Trage logging utility"
.SH NAME
TraceLog - Trace logging code instrumentation utility
.SH DESCRIPTION
The TraceLog functionality will provide a basic source code instrumentation logging
mechanism.  It will provide for a macro based API that will allow instrumentation
of the code for debug and diagnostics purposes.  The log format will provide a high
resolution timestamp, code location information (file, function, line) and an optional
user message.  The optional user message uses a varargs format that is identical to
using 'printf', with the only caveat being that a newline ("\n") is automatically
inserted, so need not be explicitley provided.

This utility is integrated with the TraceFilter mechanism to allow for full dynamic
control via the PSHELL client interface.
.SH SYNOPSIS

#include <TraceLog.h>

/* this trace cannot be disabled */
.br
#define TRACE_FORCE(format, args...)

/* normal trace output macros, these are called directly by client code */
.br
#define TRACE_ERROR(format, args...)
.br
#define TRACE_WARNING(format, args...)
.br
#define TRACE_FAILURE(format, args...)
.br
#define TRACE_INFO(format, args...)
.br
#define TRACE_DEBUG(format, args...)
.br
#define TRACE_ENTER(format, args...)
.br
#define TRACE_EXIT(format, args...)

/* hex dump trace */
.br
#define TRACE_DUMP(address, length, format, args...)

/*
.br
 * trace_registerLogFunction::
.br
 *
.br
 * typedef and function to allow a client program to register a logging
.br
 * function for message output logging, if no output function is registered
.br
 * 'printf' will be used to print out the log messages
.br
 */
.br
typedef void (*TraceLogFunction)(const char *outputString_);
.br
void trace_registerLogFunction(TraceLogFunction logFunction_);

/*
.br
 * trace_setLogPrefix:
.br
 *
.br
 * set a log name that is used as a prefix for the log type,
.br
 * if this function is not called, the prefix will default to
.br
 * "TRACE", e.g. TRACE_ERROR, TRACE_WARNING etc, if this function
.br
 * is called with a NULL prefix, the trace name type will have
.br
 * no prefix, e.g. ERROR, WARNING etc.
.br
 */
.br
void trace_setLogPrefix(const char *name_);

/*
.br
 * trace_registerLevels:
.br
 *
.br
 * register all our above trace levels with the trace filter, note
.br
 * that all levels MUST be registered before calling 'tf_init'
.br
 */
.br
void trace_registerLevels(void);

Link with the library libpshell-server
.SH EXAMPLES
See the file traceFilterDemo.c in the pshell distribution package for a complete
list of examples of using the TraceLog utility trace logging system.
.SH MT-LEVEL
Safe
.SH SEE ALSO
pshell(1), TraceFilter(3)
.SH AUTHOR
Ron Iovine  --  github.com/RonIovine/pshell


src/TraceLog.c
/*******************************************************************************
 *
 * Copyright (c) 2009, Ron Iovine, All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in the
 *       documentation and/or other materials provided with the distribution.
 *     * Neither the name of Ron Iovine nor the names of its contributors
 *       may be used to endorse or promote products derived from this software
 *       without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY Ron Iovine ''AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL Ron Iovine BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR
 * BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
 * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
 * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 *
 *******************************************************************************/

#include <stdarg.h>
#include <sys/time.h>
#include <time.h>
#include <ctype.h>
#include <string.h>
#include <pthread.h>

#include <TraceLog.h>

/* constants */

#define MAX_STRING_SIZE 256

/*
 * coding convention is leading underscore for global data,
 * trailing underscore for function arguments, and no leading
 * or trailing underscores for local stack variables
 */

/*************************
 * private "member" data
 *************************/

static TraceLogFunction _logFunction = NULL;
static char _logPrefix[MAX_STRING_SIZE] = {"TRACE"};
static pthread_mutex_t _mutex = PTHREAD_MUTEX_INITIALIZER;

/***************************************
 * private "member" function prototypes
 ***************************************/
 
static void printLine(char *line_);
static void formatHeader(const char *type_,
                         const char *file_,
                         const char *function_,
                         int line_,
                         char *outputString_);

/**************************************
 * public API "member" function bodies
 **************************************/

/******************************************************************************/
/******************************************************************************/
void trace_registerLogFunction(TraceLogFunction logFunction_)
{
  if (logFunction_ != NULL)
  {
    _logFunction = logFunction_;
  }
  else
  {
    TRACE_ERROR("NULL logFunction, not registered");
  }
}

/******************************************************************************/
/******************************************************************************/
void trace_setLogPrefix(const char *name_)
{
  if ((name_ != NULL) && (strlen(name_) > 0))
  {
    int length = strlen(name_);
    for (int i = 0; i < length; i++)
    {
      _logPrefix[i] = toupper(name_[i]);
    }
    _logPrefix[length] = '_';
    _logPrefix[length+1] = 0;
  }
  else
  {
    _logPrefix[0] = 0;
  }
}

/******************************************************************************/
/******************************************************************************/
void trace_registerLevels(void)
{
  /*
   * register all our trace levels with the trace filter service
   * format of call is "name", level, isDefault, isMaskable
   */
  tf_addLevel("ERROR", TL_ERROR, true, false);
  tf_addLevel("WARNING", TL_WARNING, true, true);
  tf_addLevel("FAILURE", TL_FAILURE, true, true);
  tf_addLevel("INFO", TL_INFO, false, true);
  tf_addLevel("DEBUG", TL_DEBUG, false, true);
  tf_addLevel("ENTER", TL_ENTER, false, true);
  tf_addLevel("EXIT", TL_EXIT, false, true);
  tf_addLevel("DUMP", TL_DUMP, false, true);
}

/******************************************************************************/
/******************************************************************************/
void trace_outputLog(const char *type_,
                     const char *file_,
                     const char *function_,
                     int line_,
                     const char *format_, ...)
{
  pthread_mutex_lock(&_mutex);
  char outputString[MAX_STRING_SIZE];
  formatHeader(type_, file_, function_, line_, outputString);
  va_list args;
  va_start(args, format_);
  vsprintf(&outputString[strlen(outputString)], format_, args);
  va_end(args);
  strcat(outputString, "\n");
  printLine(outputString);
  pthread_mutex_unlock(&_mutex);
}

/******************************************************************************/
/******************************************************************************/
void trace_outputDump(void *address_,
                      unsigned length_,
                      const char *type_,
                      const char *file_,
                      const char *function_,
                      int line_,
                      const char *format_, ...)
{
  pthread_mutex_lock(&_mutex);
  char asciiLine[80];
  const unsigned bytesPerLine = 16;
  unsigned char *bytes = (unsigned char *)address_;
  unsigned short offset = 0;
  char outputString[MAX_STRING_SIZE];
  formatHeader(type_, file_, function_, line_, outputString);
  va_list args;
  va_start(args, format_);
  vsprintf(&outputString[strlen(outputString)], format_, args);
  va_end(args);
  strcat(outputString, "\n");
  printLine(outputString);
  asciiLine[0] = 0;
  outputString[0] = 0;
  for (unsigned i = 0; i < length_; i++)
  {
    // see if we are on a full line boundry
    if ((i%bytesPerLine) == 0)
    {
      // see if we need to print out our line
      if (i > 0)
      {
        /* done with this line, add the asciii data & print it */
        sprintf(&outputString[strlen(outputString)], "  %s\n", asciiLine);
        printLine(outputString);
        // line  printed, clear them it for next time
        asciiLine[0] = 0;
        outputString[0] = 0;
      }
      // format our our offset
      sprintf(&outputString[strlen(outputString)], "  %04x  ", offset);
      offset += bytesPerLine;
    }
    // create our line of ascii data, for non-printable ascii characters just use a "."
    sprintf(&asciiLine[strlen(asciiLine)], "%c", ((isprint(bytes[i]) && (bytes[i] < 128)) ? bytes[i] : '.'));
    // format our line of hex byte
    sprintf(&outputString[strlen(outputString)], "%02x ", bytes[i]);
  }
  // done, format & print the final line & ascii data
  sprintf(&outputString[strlen(outputString)], "  %*s\n", (int)(((bytesPerLine-strlen(asciiLine))*3)+strlen(asciiLine)), asciiLine);
  printLine(outputString);
  pthread_mutex_unlock(&_mutex);
}

/**************************************
 * private "member" function bodies
 **************************************/

/******************************************************************************/
/******************************************************************************/
void formatHeader(const char *type_, const char *file_, const char *function_, int line_, char *outputString_)
{
  char timestamp[MAX_STRING_SIZE];
  struct timeval tv;
  struct tm tm;
  const char *file;
  // get timestamp
  gettimeofday(&tv, NULL);
  gmtime_r(&tv.tv_sec, &tm);
  // timestamp with date & time
  //strftime(timestamp, sizeof(timestamp), "%m/%d/%y %T", &tm);
  // timestamp with time only
  strftime(timestamp, sizeof(timestamp), "%T", &tm);
  // strip off any leading path of filename
  if ((file = strrchr(file_, '/')) != NULL)
  {
    file++;
  }
  else
  {
    file = file_;
  }
  sprintf(outputString_, "%s%-7s | %s.%ld | %s(%s):%d | ", _logPrefix, type_, timestamp, tv.tv_usec, file, function_, line_);
}

/******************************************************************************/
/******************************************************************************/
void printLine(char *line_)
{
  if (_logFunction == NULL)
  {
    printf("%s", line_);
  }
  else
  {
    (*_logFunction)(line_);
  }
}
